sympy/core/compatibility.py

```diff
@@ -45,8 +45,8 @@
     * Python 2 `__builtins__`, access with Python 3 name, `builtins`
 
 Iterator/list changes:
-    * `xrange` removed in Python 3, import `xrange` for Python 2/3 compatible
-      iterator version of range
+    * `xrange` renamed as `range` in Python 3, import `range` for Python 2/3
```

Comment:
Should this be "for Python 3 compatible..."?

Reply:
* xrange exists in Python 2
* it was removed in Python 3
* if I need the iterator form of range in Python 2 and I use xrange, the program will not run in Python 3
* if I import range from compatibility then my program will be compatible when run in Python 3

So it appears that that tweak could be made. Rather than run another set of tests for this, I will include it in my next PR if your prefer.

Just let me know.

Reply:
So this got me thinking if it was right to remove xrange. As long as we support 2.X someone might want to use xrange. So to make a program run in Py 3 that still uses 2.7 features, one would `from .compatibility import xrange`. When 2.X is dropped then one would have to purge xrange from their program and there would no longer be the option to import xrange from compatibility.

Reply:
I'm reopening this to see if anyone is thinking the same thing...or maybe I am not thinking about this correctly.

Reply:
I think that `xrange` is a builtin that cannot be imported from `compatibility` in any case.
```
Python 2.7.6 (default, Nov 23 2017, 15:49:48) 
[GCC 4.8.4] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> from sympy.core.compatibility import xrange
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: cannot import name xrange
```

Reply:
Oh. So you would need a try/except in your program anyway, making having it in compatibility pointless:

```
if pyversion > 2.X:
    xrange = range
```

I will commit this as is unless there is objection and later change the '2/3' to '3'.

Reply:
The idea is that all code should use Python 3 idioms. So all code that uses `range`/`xrange` should import `range` from compatibility and use it. In Python 3, this will do nothing. In Python 2, it will use `xrange`. So either way, it will use the iterator version, but the code will be written in a way that requires no changes once we drop Python 2 support in 2020. 

I've noticed there are other instances of this in the file as well (https://github.com/sympy/sympy/issues/7258). For instance, `unicode` is removed in Python 3, so it should not be used. And why do we still have `u_decode` when we can use `u` string literals now?